leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -108,23 +117,102 @@ public T buildResponse(final Response coapResponse) {
+                        final LwM2mResponseBuilder<T> lwm2mResponseBuilder = new LwM2mResponseBuilder<T>(coapRequest,
+                                coapResponse, client, model, observationRegistry);
+                        request.accept(lwm2mResponseBuilder);
+                        return lwm2mResponseBuilder.getResponse();
+                    }
+                });
+
+        // Store pending request to cancel it on deregistration
+        addPendingRequest(destination.getRegistrationId(), coapRequest);
 
         // Send CoAP request asynchronously
         final Endpoint endpoint = getEndpointForClient(destination);
         endpoint.sendRequest(coapRequest);
     }
 
+    public synchronized void cancelPendingRequests(String registrationId) {
+        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        if (requests != null) {
+            for (Request coapRequest : requests) {
+                coapRequest.cancel();
+                removePendingRequest(registrationId, coapRequest);
+            }
+        }
+    }
+
+    private synchronized void addPendingRequest(String registrationId, Request coapRequest) {
+        // Get or create all pending requests for this registrationId
+        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        if (requests == null) {
+            requests = new ArrayList<>();
+            pendingRequests.put(registrationId, requests);
+        }
+
+        // Add the request
+        coapRequest.addMessageObserver(new CleanerMessageObserver(registrationId, coapRequest));
+        requests.add(coapRequest);
+    }
+
+    private synchronized void removePendingRequest(String registrationId, Request coapRequest) {
+        ArrayList<Request> requests = pendingRequests.get(registrationId);
+        if (requests != null) {
+            requests.remove(coapRequest);
+        }
+    }
+
+    private class CleanerMessageObserver implements MessageObserver {
+
+        private String registrationId;
+        private Request coapRequest;
+
+        public CleanerMessageObserver(String registrationId, Request coapRequest) {
+            super();
+            this.registrationId = registrationId;
+            this.coapRequest = coapRequest;
+        }
+
+        @Override
+        public void onRetransmission() {
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            // we ignore response for observer request
+            // TODO remove this test when we will integrate the new observation registry implementation.
+            if (coapRequest.getOptions() == null
+                    || !(coapRequest.getOptions().hasObserve() && coapRequest.getOptions().getObserve() == 0)) {
+                removePendingRequest(registrationId, coapRequest);
+            }
+        }
+
+        @Override
+        public void onAcknowledgement() {
+            removePendingRequest(registrationId, coapRequest);
```

Comment:
This will "fix/bind" the usagee of piggybacked reponses.
Though "separate responses" have also advantages, I'm not sure, if binding to hard to piggybacked is a good idea.